homeassistant/components/tplink/entity.py

```diff
@@ -18,13 +24,26 @@
 _P = ParamSpec("_P")
 
 
-def async_refresh_after(
+def async_exception_wrap_refresh_after(
     func: Callable[Concatenate[_T, _P], Awaitable[None]],
 ) -> Callable[Concatenate[_T, _P], Coroutine[Any, Any, None]]:
-    """Define a wrapper to refresh after."""
+    """Define a wrapper to raise HA errors and refresh after."""
 
     async def _async_wrap(self: _T, *args: _P.args, **kwargs: _P.kwargs) -> None:
-        await func(self, *args, **kwargs)
+        try:
+            await func(self, *args, **kwargs)
+        except AuthenticationException as ex:
+            if self.coordinator.config_entry:
+                self.hass.loop.call_soon(
```

Comment:
I think it would be preferred to use `self.coordinator.config_entry.create_task()`

Reply:
It will always have a config entry

We should be able to call it directly

`self.coordinator.config_entry.async_start_reauth(self.hass)`

Reply:
We can override the config_entry type in the coordinator so it doesn't ever include `None`

Reply:
> I think it would be preferred to use `self.coordinator.config_entry.create_task()`

`create_task()` is for a coroutine, `async_start_reauth()` is not a coroutine so I think `call_soon()` is correct. (N.B. `async_start_reauth()` calls `create_task()` itself)

Reply:
> We can override the config_entry type in the coordinator so it doesn't ever include `None` because it never will

Done